### PR TITLE
External KM implementation for Forgerock

### PR DESCRIPTION
**Pull request by shilmyhasan**
This implementation allows you to integrate the WSO2 API Store with an external Identity and Access Management (IAM) by using the Forgerock OAuth Authorization Server, whihc is used to manage the OAuth clients and tokens required by WSO2 API Manager.

### CODE REVIEW

File: components/forgerock.key.manager/pom.xml [hunk 1]
```diff
@@ -0,0 +1,80 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
```
Comment (tharindu1st) [inline components/forgerock.key.manager/pom.xml:19]: licence header ?

File: components/forgerock.key.manager/src/main/java/org/wso2/forgerock/client/ForgerockConnectorConfiguration.java [hunk 2]
```diff
@@ -0,0 +1,115 @@
+/*
+ *  Copyright (c) 2020, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (tharindu1st) [inline components/forgerock.key.manager/src/main/java/org/wso2/forgerock/client/ForgerockConnectorConfiguration.java]: 2021 ?

File: components/forgerock.key.manager/src/main/java/org/wso2/forgerock/client/ForgerockConnectorConfiguration.java [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+
+        return ForgerockJWTValidatoImpl.class.getName();
+    }
+
+    @Override
+    public List<ConfigurationDto> getConnectionConfigurations() {
+
+        List<ConfigurationDto> configurationDtoList = new ArrayList<>();
+        configurationDtoList
+                .add(new ConfigurationDto("accessToken", "Registration AccessToken", "input",
```
Comment (tharindu1st) [inline components/forgerock.key.manager/src/main/java/org/wso2/forgerock/client/ForgerockConnectorConfiguration.java]: use constants for accessToken,client_id,client_secret ,etc

File: components/forgerock.key.manager/src/main/java/org/wso2/forgerock/client/ForgerockConnectorConfiguration.java [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+        configurationDtoList
+                .add(new ConfigurationDto("application_type", "Application Type", "select", "Type Of Application to " +
+                        "create", "web", false,
+                        false, Arrays.asList("web", "native", "service", "browser"), false));
+        configurationDtoList
+                .add(new ConfigurationDto("response_types", "Response Type", "select", "Type Of Token response", "",
+                        false,
+                        false, Arrays.asList("code", "token", "id_token"), true));
+        configurationDtoList
+                .add(new ConfigurationDto("access_token_lifetime", "Access Token Lifetime",
```
Comment (tharindu1st) [inline components/forgerock.key.manager/src/main/java/org/wso2/forgerock/client/ForgerockConnectorConfiguration.java]: do we have support for these from forgerock ?

File: components/forgerock.key.manager/src/main/java/org/wso2/forgerock/client/ForgerockConstants.java [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+    public static final String ERROR_ENCODING_METHOD_NOT_SUPPORTED = "Encoding method is not supported";
+    public static final String ERROR_COULD_NOT_READ_HTTP_ENTITY = "Could not read http entity for response";
+    public static final String STRING_FORMAT = "%s %s";
+    public static final String ERROR_OCCURRED_WHILE_READ_OR_CLOSE_BUFFER_READER = "Error has occurred while reading " +
+            "or closing buffer reader";
+    public static final String REQUEST_OBJECT_SIGNING_ALGORITHM = "request_object_signing_alg";
+    public static final String JWKS = "jwks";
+    public static final String SCOPE = "scp";
+    public static final String FORGEROCK_TYPE = "ForgeRock";
+    public static final String CID = "aud";
```
Comment (tharindu1st) [inline components/forgerock.key.manager/src/main/java/org/wso2/forgerock/client/ForgerockConstants.java]: CID -> AUD

File: components/forgerock.key.manager/src/main/java/org/wso2/forgerock/client/ForgerockConstants.java [hunk 6]
```diff
@@ -0,0 +1,69 @@
+/*
+/*
+ * Copyright (c) 2018, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (tharindu1st) [inline components/forgerock.key.manager/src/main/java/org/wso2/forgerock/client/ForgerockConstants.java]: 2021

File: components/forgerock.key.manager/src/main/java/org/wso2/forgerock/client/ForgerockJWTValidatoImpl.java [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+import org.wso2.carbon.apimgt.api.APIManagementException;
+import org.wso2.carbon.apimgt.impl.dto.JWTValidationInfo;
+import org.wso2.carbon.apimgt.impl.jwt.JWTValidatorImpl;
+import org.wso2.carbon.apimgt.impl.jwt.SignedJWTInfo;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+public class ForgerockJWTValidatoImpl  extends JWTValidatorImpl {
```
Comment (tharindu1st) [inline components/forgerock.key.manager/src/main/java/org/wso2/forgerock/client/ForgerockJWTValidatoImpl.java:29]: class comment ?

File: components/forgerock.key.manager/src/main/java/org/wso2/forgerock/client/ForgerockOAuthClient.java [hunk 8]
```diff
@@ -0,0 +1,748 @@
+/*
+ * Copyright (c) 2018, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (tharindu1st) [inline components/forgerock.key.manager/src/main/java/org/wso2/forgerock/client/ForgerockOAuthClient.java]: 2021

**Review by tharindu1st — CHANGES_REQUESTED** [review]

**Comment by tharindu1st** [comment]
please add check-style and findbugs plugins to the repo.

File: components/forgerock.key.manager/src/main/java/org/wso2/forgerock/client/ForgerockConnectorConfiguration.java (hunk 4 above)
Comment (shilmyhasan) [reply components/forgerock.key.manager/src/main/java/org/wso2/forgerock/client/ForgerockConnectorConfiguration.java]: Yeah

File: components/forgerock.key.manager/pom.xml (hunk 1 above)
Comment (shilmyhasan) [reply components/forgerock.key.manager/pom.xml:19]: Fixed in 0cbcfc879bd87c85a4c14d959c2d28653f7fd391

File: components/forgerock.key.manager/src/main/java/org/wso2/forgerock/client/ForgerockConnectorConfiguration.java (hunk 2 above)
Comment (shilmyhasan) [reply components/forgerock.key.manager/src/main/java/org/wso2/forgerock/client/ForgerockConnectorConfiguration.java]: Fixed in 0cbcfc879bd87c85a4c14d959c2d28653f7fd391

File: components/forgerock.key.manager/src/main/java/org/wso2/forgerock/client/ForgerockConnectorConfiguration.java (hunk 3 above)
Comment (shilmyhasan) [reply components/forgerock.key.manager/src/main/java/org/wso2/forgerock/client/ForgerockConnectorConfiguration.java]: Fixed in 0cbcfc879bd87c85a4c14d959c2d28653f7fd391

File: components/forgerock.key.manager/src/main/java/org/wso2/forgerock/client/ForgerockConstants.java (hunk 5 above)
Comment (shilmyhasan) [reply components/forgerock.key.manager/src/main/java/org/wso2/forgerock/client/ForgerockConstants.java]: Fixed in 0cbcfc879bd87c85a4c14d959c2d28653f7fd391

File: components/forgerock.key.manager/src/main/java/org/wso2/forgerock/client/ForgerockConstants.java (hunk 6 above)
Comment (shilmyhasan) [reply components/forgerock.key.manager/src/main/java/org/wso2/forgerock/client/ForgerockConstants.java]: Fixed in 0cbcfc879bd87c85a4c14d959c2d28653f7fd391

File: components/forgerock.key.manager/src/main/java/org/wso2/forgerock/client/ForgerockJWTValidatoImpl.java (hunk 7 above)
Comment (shilmyhasan) [reply components/forgerock.key.manager/src/main/java/org/wso2/forgerock/client/ForgerockJWTValidatoImpl.java:29]: Added in 0cbcfc879bd87c85a4c14d959c2d28653f7fd391

File: components/forgerock.key.manager/src/main/java/org/wso2/forgerock/client/ForgerockOAuthClient.java (hunk 8 above)
Comment (shilmyhasan) [reply components/forgerock.key.manager/src/main/java/org/wso2/forgerock/client/ForgerockOAuthClient.java]: Fixed in 0cbcfc879bd87c85a4c14d959c2d28653f7fd391

File: components/forgerock.key.manager/src/main/java/org/wso2/forgerock/client/dao/ForgerockApiMgtDAO.java [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+
+/**
+ * This class provides the DAO implementation for "Forgerock" connector
+ */
+public class ForgerockApiMgtDAO {
+
+    private static ForgerockApiMgtDAO INSTANCE = null;
```
Comment (tharindu1st) [inline components/forgerock.key.manager/src/main/java/org/wso2/forgerock/client/dao/ForgerockApiMgtDAO.java]: private static ForgerockApiMgtDAO INSTANCE = new ForgerockApiMgtDAO()'

File: components/forgerock.key.manager/src/main/java/org/wso2/forgerock/client/dao/ForgerockApiMgtDAO.java [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+    private static final Log log = LogFactory.getLog(ForgerockApiMgtDAO.class);
+
+    /**
+     * Method to get the instance of the ApiMgtDAO.
+     *
+     * @return {@link ForgerockApiMgtDAO} instance
+     */
+    public static ForgerockApiMgtDAO getInstance() {
+        if (INSTANCE == null) {
```
Comment (tharindu1st) [inline components/forgerock.key.manager/src/main/java/org/wso2/forgerock/client/dao/ForgerockApiMgtDAO.java]: remove this if condition and return INSTANCE

File: components/forgerock.key.manager/src/main/java/org/wso2/forgerock/client/dao/ForgerockApiMgtDAO.java (hunk 9 above)
Comment (tharindu1st) [reply components/forgerock.key.manager/src/main/java/org/wso2/forgerock/client/dao/ForgerockApiMgtDAO.java]: Add Private ForgerockApiMgtDAO constructor to not to create new

File: components/forgerock.key.manager/src/main/java/org/wso2/forgerock/client/ForgerockOAuthClient.java [hunk 11]
```diff
@@ -0,0 +1,828 @@
...
+/**
+ * This class provides the implementation to use "Forgerock" for managing
+ * OAuth clients and Tokens needed by WSO2 API Manager.
+ */
+public class ForgerockOAuthClient extends AbstractKeyManager {
+
+    private static final Log log = LogFactory.getLog(ForgerockOAuthClient.class);
+    private ForgerockDCRClient forgeDCRClient;
+    private IntrospectClient introspectClient;
+    ForgerockApiMgtDAO apiMgtDAO = ForgerockApiMgtDAO.getInstance();
```
Comment (tharindu1st) [inline components/forgerock.key.manager/src/main/java/org/wso2/forgerock/client/ForgerockOAuthClient.java:65]: forgerockDao

File: components/forgerock.key.manager/src/main/java/org/wso2/forgerock/client/ForgerockOAuthClient.java [hunk 12]
```diff
@@ -0,0 +1,828 @@
...
+        ClientInfo retrievedClientInfo = forgeDCRClient.getApplication(clientId);
+        return createOAuthAppInfoFromResponse(retrievedClientInfo);
+    }
+
+    /**
+     * Gets new access token and returns it in an AccessTokenInfo object.
+     * @return AccessTokenInfo Info of the new token.
+     * @throws APIManagementException This is the custom exception class for API management.
+     */
+    public String getRegistrationAccessToken()
```
Comment (tharindu1st) [inline components/forgerock.key.manager/src/main/java/org/wso2/forgerock/client/ForgerockOAuthClient.java:201]: public-> private

**Review by tharindu1st — CHANGES_REQUESTED** [review]

**Review by tharindu1st — APPROVED** [review]

**Pull request merged**